src/appleseed/renderer/modeling/object/rectobject.cpp

```diff
@@ -0,0 +1,310 @@
+//
+
+// interface header.
+#include "rectobject.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/shading/shadingray.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/intersection/rayparallelogram.h"
+#include "foundation/math/ray.h"
+#include "foundation/math/scalar.h"
+#include "foundation/math/vector.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/containers/dictionary.h"
+#include "foundation/utility/job/iabortswitch.h"
+#include "foundation/utility/searchpaths.h"
+#include "foundation/utility/string.h"
+
+using namespace foundation;
+
+namespace renderer
+{
+
+//
+// RectObject class implementation.
+//
+
+namespace
+{
+    const char* Model = "rect_object";
+}
+
+RectObject::RectObject(
+    const char*            name,
+    const ParamArray&      params)
+  : ProceduralObject(name, params)
+{
+}
+
+void RectObject::release()
+{
+    delete this;
+}
+
+const char* RectObject::get_model() const
+{
+    return Model;
+}
+
+bool RectObject::on_frame_begin(
+    const Project&         project,
+    const BaseGroup*       parent,
+    OnFrameBeginRecorder&  recorder,
+    IAbortSwitch*          abort_switch)
+{
+    if (!ProceduralObject::on_frame_begin(project, parent, recorder, abort_switch))
+        return false;
+
+    const double width  = get_uncached_width();
+    const double height = get_uncached_height();
+
+    m_half_width = width * 0.5;
+    m_half_height = height * 0.5;
+
+    if (width != 0.0)
+        m_rcp_width = 1.0 / width;
+
+    if (height != 0.0)
+        m_rcp_height = 1.0 / height;
+
+    return true;
+}
+
+GAABB3 RectObject::compute_local_bbox() const
+{
+    const float width  = static_cast<float>(get_uncached_width());
+    const float height = static_cast<float>(get_uncached_height());
+
+    const GVector3 pmax(width * 0.5f, 0.0f, height * 0.5f);
+    const GVector3 pmin(-pmax.x, 0.0f, -pmax.z);
+
+    GAABB3 bbox(pmin, pmax);
+    return bbox;
+}
+
+size_t RectObject::get_material_slot_count() const
+{
+    return 1;
+}
+
+const char* RectObject::get_material_slot(const size_t index) const
+{
+    return "default";
+}
+
+double RectObject::get_uncached_width() const
+{
+    return m_params.get_optional<double>("width", 1.0);
+}
+
+double RectObject::get_uncached_height() const
+{
+    return m_params.get_optional<double>("height", 1.0);
+}
+
+void RectObject::get_origin_and_axes(
+    Vector3d&              origin,
+    Vector3d&              x,
+    Vector3d&              y,
+    Vector3d&              n) const
+{
+    const double width  = get_uncached_width();
+    const double height = get_uncached_height();
+
+    origin = Vector3d(-width * 0.5, 0.0, -height * 0.5);
+    x = Vector3d(width, 0.0, 0.0);
+    y = Vector3d(0.0, 0.0, height);
+    n = Vector3d(0.0, 1.0, 0.0);
+}
+
+void RectObject::intersect(
+    const ShadingRay&      ray,
+    IntersectionResult&    result) const
+{
+    if (m_half_width == 0.0 || m_half_height == 0.0)
```

Comment:
@est77 Isn't it better to assume that it's not 0 ?

Reply:
0 width or height are valid values. The only thing you need is to make sure that a ray will never intersect the rect.
Speaking of unlikely events, this looks like a good place to use APPLESEED_UNLIKELY()

Reply:
Something that we should do maybe is to prevent negative widths and heights. clamping to 0.

Reply:
We can replace the inline intersection code here by the ray parallelogram intersection function.
The rect corner that we need can be precomputed.

Reply:
Addressed. We don't need to clamp with and height, the metadata allows [0, inf)